server/src/main/java/org/opensearch/common/cache/service/NodeCacheStats.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.cache.service;
+
+import org.opensearch.action.admin.indices.stats.CommonStatsFlags;
+import org.opensearch.common.annotation.ExperimentalApi;
+import org.opensearch.common.cache.CacheType;
+import org.opensearch.common.cache.stats.ImmutableCacheStatsHolder;
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.common.io.stream.StreamOutput;
+import org.opensearch.core.common.io.stream.Writeable;
+import org.opensearch.core.xcontent.ToXContentFragment;
+import org.opensearch.core.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+
+/**
+ * A class creating XContent responses to cache stats API requests.
+ *
+ * @opensearch.experimental
+ */
+@ExperimentalApi
+public class NodeCacheStats implements ToXContentFragment, Writeable {
+    private final TreeMap<CacheType, ImmutableCacheStatsHolder> statsByCache;
+    private final CommonStatsFlags flags;
+
+    public NodeCacheStats(Map<CacheType, ImmutableCacheStatsHolder> statsByCache, CommonStatsFlags flags) {
+        this.statsByCache = new TreeMap<>(statsByCache); // Use TreeMap to force consistent ordering of caches in API responses
+        this.flags = flags;
+    }
+
+    public NodeCacheStats(StreamInput in) throws IOException {
+        this.flags = new CommonStatsFlags(in);
+        Map<CacheType, ImmutableCacheStatsHolder> readMap = in.readMap(i -> i.readEnum(CacheType.class), ImmutableCacheStatsHolder::new);
+        this.statsByCache = new TreeMap<>(readMap);
```

Comment:
It might be tricky to ser/deser a TreeMap in general because users might want to supply their own Comparator for the ordering. (In our case we use the default alphabetic ordering so it's not a concern). 

The caller of the serializer and the deserializer probably shouldn't both have to have access to the comparator object to correctly rebuild the map, so it would have to be serialized along with the map somehow, but I'm not sure how this would be achieved.

Reply:
I am wondering if anything is stopping us from refactoring it like this (we can take comparator as input if don't want to use natural comparator?):

```
/**
     * If the returned map contains any entries it will be mutable. If it is empty it might be immutable.
     */
    public <K, V> Map<K, V> readMap(Writeable.Reader<K> keyReader, Writeable.Reader<V> valueReader) throws IOException {
        int size = readArraySize();
        if (size == 0) {
            return Collections.emptyMap();
        }
        Map<K, V> map = new HashMap<>(size);
        readIntoMap(keyReader, valueReader, map, size);
        return map;
    }

    public <K, V> SortedMap<K, V> readOrderedMap(Writeable.Reader<K> keyReader, Writeable.Reader<V> valueReader) throws IOException {
        int size = readArraySize();
        if (size == 0) {
            return Collections.emptySortedMap();
        }
        SortedMap<K, V> sortedMap = new TreeMap<>();
        readIntoMap(keyReader, valueReader, sortedMap, size);
        return sortedMap;
    }

    private <K, V> void readIntoMap(Writeable.Reader<K> keyReader, Writeable.Reader<V> valueReader, Map<K, V> map, int size) throws IOException {
        for (int i = 0; i < size; i++) {
            K key = keyReader.read(this);
            V value = valueReader.read(this);
            map.put(key, value);
        }
    }
```

Reply:
Yeah, this is fine if we're ok with providing a comparator during serialization/deserialization